Make it compatible with both python2 and python3

jon1012 opened the pull request:
Right now it depends on past/future from futurize. We can maybe avoid this dependency ?

Commit:
using setuptools

Commit:
Porting to Python 3

Commit:
Adding ply as a dependency

Commit:
woops, should be install_requires

Commit:
install_requireS even.

timtadh commented:
@jon1012 not an expert on py2/3 compat. Also if you want to use the repl you need to depend the getline library as well. It is in the `reqs.txt` file.

I will try and give this a test next week.

timtadh commented:
Had tons of errors when I ran the pr in python 2. I haven't tested with python 3. Also, `future` needs to be added to both the reqs and the setup.py file. I think I can push to the pr. If not can you add that?

```
test_DEC (pyflwor.test_lexer.TestLexer) ... ok
test_FLOAT (pyflwor.test_lexer.TestLexer) ... ok
test_HEX (pyflwor.test_lexer.TestLexer) ... ok
test_KEYWORDS (pyflwor.test_lexer.TestLexer) ... ok
test_NAME (pyflwor.test_lexer.TestLexer) ... ok
test_OCT (pyflwor.test_lexer.TestLexer) ... ok
test_STRING (pyflwor.test_lexer.TestLexer) ... ok
test_chrs (pyflwor.test_lexer.TestLexer) ... ok
test_contextmng (pyflwor.test_lexer.TestLexer) ... ok
test_flwr (pyflwor.test_parser.TestParser) ... ERROR
test_flwr_attrvalue (pyflwor.test_parser.TestParser) ... ERROR
test_flwr_function_args (pyflwor.test_parser.TestParser) ... ERROR
test_flwr_function_noargs (pyflwor.test_parser.TestParser) ... ERROR
test_flwr_orderby (pyflwor.test_parser.TestParser) ... ERROR
test_hello (pyflwor.test_parser.TestParser) ... ERROR
test_if (pyflwor.test_parser.TestParser) ... ERROR
test_reduce (pyflwor.test_parser.TestParser) ... ERROR
test_setops (pyflwor.test_parser.TestParser) ... ERROR
test_slash (pyflwor.test_parser.TestParser) ... ERROR
test_where_bool (pyflwor.test_parser.TestParser) ... ERROR
test_where_exists (pyflwor.test_parser.TestParser) ... ERROR
test_where_op (pyflwor.test_parser.TestParser) ... ERROR
test_where_parens_bool (pyflwor.test_parser.TestParser) ... ERROR
test_where_quantified (pyflwor.test_parser.TestParser) ... ERROR
test_where_setcmp (pyflwor.test_parser.TestParser) ... ERROR
test_where_value (pyflwor.test_parser.TestParser) ... ERROR
test_where_value_params (pyflwor.test_parser.TestParser) ... ERROR
test_None (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_arithmetic (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_callable_where_values (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_cmpops (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_construct_class (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_count (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_dict_where_values (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_flattened_return (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_flwr_attrvalue (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_flwr_orderby (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_func_where_values (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_function_def (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_hello (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_ifExpr (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_ifExpr_short_circuit (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_innotin (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_iterdown (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_list_literal (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_list_where_values (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_multi_collect (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_nested_boolean_exprs (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_no_for (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_simple_where_values (pyflwor.test_pyflwor.TestPyQuery) ... ERROR
test_smpl_boolean_exprs (pyflwor.test_pyflwor.TestPyQuery) ... ERROR

======================================================================
ERROR: test_flwr (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 112, in test_flwr
    pyflwor.compile('for x in <asfd> return x')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_flwr_attrvalue (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 172, in test_flwr_attrvalue
    return 'asdf':asdf, "one":1, "2.0":2.0''')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_flwr_function_args (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 208, in test_flwr_function_args
    ''')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_flwr_function_noargs (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 201, in test_flwr_function_noargs
    ''')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_flwr_orderby (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 181, in test_flwr_orderby
    pyflwor.compile('for x in <asdf> order by "adsf" desc return "adsf":x')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_hello (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 22, in test_hello
    pyflwor.compile('hello')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_if (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 213, in test_if
    ''')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_reduce (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 221, in test_reduce
    ''')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_setops (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 105, in test_setops
    pyflwor.compile('asdf - asdf')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_slash (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 26, in test_slash
    pyflwor.compile('hello/hello')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_where_bool (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 48, in test_where_bool
    pyflwor.compile('hello[a and b]')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_where_exists (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 33, in test_where_exists
    pyflwor.compile('hello[wheRe]/hello[asdf]')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_where_op (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 40, in test_where_op
    pyflwor.compile('hello[1 == 1]')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_where_parens_bool (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 58, in test_where_parens_bool
    pyflwor.compile('hello[((a and b) and not (a or b) or not (a and b)) and not (not a or b)]')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_where_quantified (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 84, in test_where_quantified
    pyflwor.compile('hello[every x in <asdf> satisfies (x)]')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_where_setcmp (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 94, in test_where_setcmp
    pyflwor.compile('hello[a in <qs>]')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_where_value (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 61, in test_where_value
    pyflwor.compile('hello[a]')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_where_value_params (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 76, in test_where_value_params
    pyflwor.compile('hello[f(1)]')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_None (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 286, in test_None
    ''', d), oset([None]))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_arithmetic (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 358, in test_arithmetic
    ''', d), (
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_callable_where_values (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 171, in test_callable_where_values
    self.assertEquals(exe('a[m["next"][7](j)(m["next"][7])(m["next"])[7](i)(m["two"]) == 4]', d), oset([a]))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_cmpops (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 47, in test_cmpops
    self.assertEquals(exe('a[self.q == 5]', locals()), oset([a]))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_construct_class (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 298, in test_construct_class
    ''', d))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_count (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 331, in test_count
    ''', d), {1:1,2:1,3:3,4:3,5:2,6:2,7:2})
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_dict_where_values (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 151, in test_dict_where_values
    self.assertEquals(exe('a[l["one"] == 1]', d), oset([a]))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_flattened_return (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 264, in test_flattened_return
    ''', d), (1, 2, 3, 4, 5, 6, 7, 1, 2, 3, 4, 5, 6, 7, 1, 2, 3, 4, 5, 6, 7, 8))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_flwr_attrvalue (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 178, in test_flwr_attrvalue
    self.assertEquals(exe('for x in f() return x', d), (1,2,3))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_flwr_orderby (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 186, in test_flwr_orderby
    self.assertRaises(SyntaxError, exe, 'for x in f() order by "asdf" ascd return x', d)
  File "/usr/lib/python2.7/unittest/case.py", line 475, in assertRaises
    callableObj(*args, **kwargs)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_func_where_values (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 125, in test_func_where_values
    self.assertEquals(exe('a[f()]', d), oset([a]))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_function_def (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 201, in test_function_def
    ''', d), (125, 125, 125, 125, 125, 125, 125, 125))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_hello (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 27, in test_hello
    q = pyflwor.compile('hello')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_ifExpr (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 231, in test_ifExpr
    ''', d), oset([int(q)]))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_ifExpr_short_circuit (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 243, in test_ifExpr_short_circuit
    ''', d), (1,))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_innotin (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 378, in test_innotin
    ''', d), {1:1,2:1,3:3,4:3,5:2,6:2,7:2})
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_iterdown (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 39, in test_iterdown
    self.assertEquals(exe('o/x[hasattr(self,"y")]/y/q', d), oset([answer]))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_list_literal (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 342, in test_list_literal
    ''', d), {1:1,2:1,3:3,4:3,5:2,6:2,7:2})
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_list_where_values (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 137, in test_list_where_values
    self.assertEquals(exe('a[l[0] == 1]', d), oset([a]))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_multi_collect (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 393, in test_multi_collect
    ''', d), (
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_nested_boolean_exprs (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 91, in test_nested_boolean_exprs
    self.assertEquals(exe('a[true and (false or true)]', locals()), oset([a]))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_no_for (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 311, in test_no_for
    ''', d), (l,))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_simple_where_values (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 105, in test_simple_where_values
    self.assertEquals(exe('a[1 == 1]', d), oset([a]))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

======================================================================
ERROR: test_smpl_boolean_exprs (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 70, in test_smpl_boolean_exprs
    self.assertEquals(exe('a[true]', locals()), oset([a]))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 29, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 22, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
TypeError: str() takes at most 1 argument (2 given)

----------------------------------------------------------------------
Ran 51 tests in 0.159s

FAILED (errors=42)
```

timtadh commented:
@jon1012 looks like I can't update your pr. Can you apply the patch?

```
From e726adea06144b47984d3340608f82ebcc750da2 Mon Sep 17 00:00:00 2001
From: Tim Henderson <tim.tadh@gmail.com>
Date: Mon, 23 Mar 2015 08:57:54 -0400
Subject: [PATCH] Added future to the requirements

Signed-off-by: Tim Henderson <tim.tadh@gmail.com>
---
 reqs.txt | 1 +
 setup.py | 2 +-
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/reqs.txt b/reqs.txt
index f4c139d..f694cdb 100644
--- a/reqs.txt
+++ b/reqs.txt
@@ -1,3 +1,4 @@
 -egit+https://github.com/timtadh/getline.git#egg=getline
 nose
 ply
+future
diff --git a/setup.py b/setup.py
index 5b940d2..fd50cfd 100644
--- a/setup.py
+++ b/setup.py
@@ -11,6 +11,6 @@ setup(name='pyflwor',
       url='https://www.github.com/timtadh/pyflwor',
       license='BSD',
       packages=['pyflwor'],
-      install_requires=['ply']
+      install_requires=['ply', 'future']
 )

-- 
1.9.1
```

Commit:
adding future/builtins reqs for python2 compat

Commit:
adding future as a requirement

jon1012 commented:
Ok, I added this requirement and added  a fix for the probem you got. Now all tests pass on python2 :)

I just gave you write access to the branch too.

jon1012 commented:
Hi,

Any news on that pull request ?

Jonathan

timtadh commented:
Hi @jon1012 

Sorry this dropped off my plate. I just tested this. I got the following errors in 3.4:

```
======================================================================
ERROR: test_where_op (pyflwor.test_parser.TestParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_parser.py", line 43, in test_where_op
    pyflwor.compile('hello[1 <= 1]')
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 24, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
  File "/home/hendersont/stuff/code/pyflwor/env3/lib/python3.4/site-packages/ply/yacc.py", line 331, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/home/hendersont/stuff/code/pyflwor/env3/lib/python3.4/site-packages/ply/yacc.py", line 1181, in parseopt_notrack
    tok = call_errorfunc(self.errorfunc, errtoken, self)
  File "/home/hendersont/stuff/code/pyflwor/env3/lib/python3.4/site-packages/ply/yacc.py", line 193, in call_errorfunc
    r = errorfunc(token)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/parser.py", line 597, in p_error
    raise SyntaxError("Syntax error at '%s', %s.%s" % (t,t.lineno,t.lexpos))
  File "<string>", line None
SyntaxError: Syntax error at 'LexToken(EQ,'=',1,9)', 1.9

======================================================================
ERROR: test_cmpops (pyflwor.test_pyflwor.TestPyQuery)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_pyflwor.py", line 50, in test_cmpops
    self.assertEquals(exe('a[self.q <= 5]', locals()), oset([a]))
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 31, in execute
    return compile(query)(namespace)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/pyflwor.py", line 24, in compile
    return Parser().parse(bytes(query, 'utf-8').decode('unicode_escape'), lexer=Lexer())
  File "/home/hendersont/stuff/code/pyflwor/env3/lib/python3.4/site-packages/ply/yacc.py", line 331, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/home/hendersont/stuff/code/pyflwor/env3/lib/python3.4/site-packages/ply/yacc.py", line 1181, in parseopt_notrack
    tok = call_errorfunc(self.errorfunc, errtoken, self)
  File "/home/hendersont/stuff/code/pyflwor/env3/lib/python3.4/site-packages/ply/yacc.py", line 193, in call_errorfunc
    r = errorfunc(token)
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/parser.py", line 597, in p_error
    raise SyntaxError("Syntax error at '%s', %s.%s" % (t,t.lineno,t.lexpos))
  File "<string>", line None
SyntaxError: Syntax error at 'LexToken(EQ,'=',1,10)', 1.10

======================================================================
FAIL: test_chrs (pyflwor.test_lexer.TestLexer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_lexer.py", line 125, in test_chrs
    for t2 in tokens: self.assertEquals(next(clex), t2)
AssertionError: LexToken(LANGLE,'<',1,0) != LexToken(LE,'<=',1,0)

======================================================================
FAIL: test_contextmng (pyflwor.test_lexer.TestLexer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/hendersont/stuff/code/pyflwor/pyflwor/test_lexer.py", line 50, in test_contextmng
    self.assertNotEquals(t1, t2)
AssertionError: LexToken(NAME,'a',1,0) == LexToken(NAME,'a',1,0)

----------------------------------------------------------------------
```

Furthermore, it seems to regenerate the LALR parsing tables everytime. Not sure why that happens. So if you can get those two things fixed plus add the following "reqs3.txt" file I am happy to merge.

```
future
nose
ply
```

Commit:
fixing relative imports

Commit:
fixing orderedset import

jon1012 committed:
removing debug and adding optimize

jon1012 committed:
trying some other optim

Commit:
forced version number

timtadh commented:
Hi @jon1012 sorry I forgot to close this so long ago. I just got pinged by your new commits. It looks fine to me. Do you want me to add you as a maintainer on this repo?
